File: Difficulty: Medium/Word Search/word-search.js
//{ Driver Code Starts
"use strict";

process.stdin.resume();
process.stdin.setEncoding("utf-8");

let inputString = "";
let currentLine = 0;

process.stdin.on("data", (inputStdin) => { inputString += inputStdin; });

process.stdin.on("end", (_) => {
    inputString =
        inputString.trim().split("\n").map((string) => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {
        let n = parseInt(readLine());
        let m = parseInt(readLine());
        let mat = new Array(n);

        for (let j = 0; j < n; j++) {
            let input_line = readLine().split(" ");
            mat[j] = new Array(m);
            for (let k = 0; k < m; k++) mat[j][k] = input_line[k];
        }

        let word = readLine();
        let obj = new Solution();
        let res = obj.isWordExist(mat, word);

        if (res) {
            console.log("true");
        } else {
            console.log("false");
        }

        console.log("~");
    }
}
// } Driver Code Ends


// User function Template for javascript

/**
 * @param {string[]} mat
 * @param {string} word
 * @returns {boolean}
 */

class Solution {
    isWordExist(mat, word) {
        const n = mat.length;
        const m = mat[0].length;
        const directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]; // directions for up, down, left, right

        // Helper function for Depth First Search (DFS)
        function dfs(i, j, index) {
            // If the index equals the length of the word, we've found the word
            if (index === word.length) return true;

            // Check for out of bounds or mismatch
            if (i < 0 || i >= n || j < 0 || j >= m || mat[i][j] !== word[index]) {
                return false;
            }

            // Temporarily mark this cell as visited by replacing it with a special character (e.g., '#')
            const temp = mat[i][j];
            mat[i][j] = '#';

            // Explore all four possible directions
            for (let [dx, dy] of directions) {
                const ni = i + dx;
                const nj = j + dy;
                if (dfs(ni, nj, index + 1)) {
                    return true;
                }
            }

            // Backtrack: restore the original value of the cell
            mat[i][j] = temp;
            return false;
        }

        // Start from every cell in the matrix
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < m; j++) {
                if (mat[i][j] === word[0] && dfs(i, j, 0)) {
                    return true;
                }
            }
        }

        return false;
    }
}

File: Difficulty: Medium/Word Search/word-search.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.math.*;
import java.io.*;

class GFG {
  public static void main(String[] args) throws IOException {
    Scanner sc = new Scanner(System.in);
    int T = sc.nextInt();
    while (T-- > 0) {
      int n = sc.nextInt();
      int m = sc.nextInt();
      char a[][] = new char[n][m];
      sc.nextLine();
      for(int i=0;i<n;i++){
        for(int j=0;j<m;j++)
         a[i][j]=sc.next().charAt(0);
      }
      String word=sc.next();
      Solution obj = new Solution();
      int ans = obj.wordSearch(a,word) ? 1 : 0;
      System.out.println(ans);
    
System.out.println("~");
}
  }
}

// } Driver Code Ends

class Solution {
  
  /**
   * Function to search for the given word in the matrix.
   * The word can be formed by sequentially moving up, down, left, or right.
   * @param m - 2D matrix of characters.
   * @param w - The word to search for.
   * @returns boolean - Returns true if the word is found, false otherwise.
   */
  public static boolean wordSearch(char m[][], String w) {
    // Iterate over each cell in the matrix
    for (int i = 0; i < m.length; i++) {
      for (int j = 0; j < m[0].length; j++) {
        // Start checking if the word can be formed starting from each cell
        if (check(m, w, i, j, 0)) {
          return true;  // Word found
        }
      }
    }
    return false;  // Word not found
  }

  /**
   * Helper function to recursively search the word in all possible directions.
   * @param m - 2D matrix of characters.
   * @param w - The word to search for.
   * @param i - Current row index.
   * @param j - Current column index.
   * @param k - Current character index in the word.
   * @returns boolean - Returns true if the word can be formed starting from current position.
   */
  static boolean check(char m[][], String w, int i, int j, int k) {
    // If all characters of the word have been matched
    if (k == w.length()) {
      return true;
    }

    // Check if we are out of bounds or the cell is already visited (marked with '#')
    if (i < 0 || j < 0 || i >= m.length || j >= m[0].length || m[i][j] == '#') {
      return false;
    }

    // If the current cell matches the character in the word
    if (m[i][j] == w.charAt(k)) {
      char temp = m[i][j];  // Store the character
      m[i][j] = '#';  // Mark the current cell as visited
      
      // Recursively check in all four directions (up, down, left, right)
      if (check(m, w, i + 1, j, k + 1) || check(m, w, i - 1, j, k + 1) ||
          check(m, w, i, j + 1, k + 1) || check(m, w, i, j - 1, k + 1)) {
        return true;  // Word found
      }

      m[i][j] = temp;  // Backtrack and restore the character
    }

    return false;  // Word not found
  }
}
